### PR TITLE
Fixes #5062

### LINKED ISSUE #5062: Enum with holes broken in JS
Enum types with holes don't seem work with the js target.

Consider the file `test.nim`:
```nim
type Holed = enum
  hFirst = 0
  hSecond = 32
  hThird = 64
  
var x = [0,32,64] # This is just to avoid the compiler inlining the value of the enum

echo Holed(x[0])
echo Holed(x[1])
echo Holed(x[2])
```
When compiled with `js `, only the first `echo` works, the others give `Unable to get property 'name' of undefined or null reference`. Compiling with `c` works as expected.

The relevant ( I think) part of the compiled `test.js`:
```js
// The enum as compiled
var NNI23004 = {kind: 2, offset: 0, typ: null, name: null, len: 3, sons: [{kind: 1, offset: 0, typ: NTI23004, name: "hFirst", len: 0, sons: null}, 
{kind: 1, offset: 32, typ: NTI23004, name: "hSecond", len: 0, sons: null}, 
{kind: 1, offset: 64, typ: NTI23004, name: "hThird", len: 0, sons: null}]};

/*...*/

rawEcho(cstrToNimstr(NTI23004.node.sons[x_23012[0][0]].name));
rawEcho(cstrToNimstr(NTI23004.node.sons[x_23012[0][1]].name));
rawEcho(cstrToNimstr(NTI23004.node.sons[x_23012[0][2]].name));
```
It seems like it's assumed that the enums are without holes and start at zero, in fact eg `x_23012[0][1]` equals 32, so it looks for element 32 in the `sons` array, going out of bounds.

A naive (slow) fix would be to check if the offset corresponds to the value, eg:
```js
function getEnum(sons,offset){
  for (i in sons){    
    if (sons[i].offset === offset) return sons[i]
  }
}
```
And generate something like:
```js
rawEcho(cstrToNimstr( getEnum(NTI23004.node.sons,x_23012[0][1]).name));
```
But there are probably better solutions.

### PR DESCRIPTION
**Pull request by stisa**
Change enums from js arrays to js object, using the offset of each element as index.  
This fixes  #5062, I also added some basic tests for enums with holes and offsets.

The behaviour should be the same, as non-holed enums simply have offsets starting from 0, but
the set of tests for the js target is a bit small, so I'd like some more validation.

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -1634,16 +1634,8 @@ proc genRepr(p: PProc, n: PNode, r: var TCompRes) =
   of tyEnum, tyOrdinal:
     gen(p, n.sons[1], r)
     useMagic(p, "cstrToNimstr")
-    var offset = ""
-    if t.kind == tyEnum:
```
Comment (Araq) [inline compiler/jsgen.nim]: Well that code wasn't added without purpose. Check the git log to see the reason for this way of doing it.

Comment (stisa) [inline compiler/jsgen.nim]: You are referring to https://github.com/nim-lang/Nim/commit/a76b74ed74c98eff41752fc9cb71676f527bab21 right?  
You are right, this breaks when the enum offset is negative.  
Wrapping the key in quotes (`\"$1\"` ) to make it a jstring should be enough to fix this. JS object key are internally converted to strings anyway.

**Comment by Araq** [comment]
@yglukhov Please review.

**Comment by yglukhov** [comment]
LGTM. For the future I would suggest using `doAssert`s in the tests because they are more maintainable and readable than `output`.

**Comment by dom96** [comment]
Please follow the style guide for code next time (even for tests). The ``type`` should be on its own line.

**Comment by stisa** [comment]
@dom96  will do, but could you link me this style guide?  
I tried looking for something like a "contributing to the compiler" guide but didn't find it, and I may have missed it but I can't find this rule about type in nep1.  
Are there other style guides?
I'm trying to get repr working in js, so I'll fix them if/when I make another pr.

**Comment by dom96** [comment]
Looks like the style guide itself (NEP1) shows

```
type PathComponent = enum
  pcDir
  pcLinkToDir
  pcFile
  pcLinkToFile
```

This is wrong and should be corrected. In fact, the wiki shows it correctly: https://github.com/nim-lang/Nim/wiki/Style-Guide-for-Nim-Code/d8a3935bc507a954032fc0458020a292236bd77f. The NEP1 document was imported from an older version of the wiki...

**Pull request merged**